Solutions/Corvus.Json.ExtendedTypes/Corvus.Json/JsonAny.cs

```diff
@@ -930,49 +965,144 @@ public void WriteTo(Utf8JsonWriter writer)
         public JsonObjectEnumerator EnumerateObject()
         {
-            return this.AsObject.EnumerateObject();
+            if (this.objectBacking is ImmutableDictionary<string, JsonAny> properties)
+            {
+                return new JsonObjectEnumerator(properties);
+            }
+
+            if (this.jsonElementBacking.ValueKind == JsonValueKind.Object)
+            {
+                return new JsonObjectEnumerator(this.jsonElementBacking);
+            }
+
+            return default;
         }
 
         /// <inheritdoc/>
         public JsonArrayEnumerator EnumerateArray()
         {
-            return this.AsArray.EnumerateArray();
+            if (this.arrayBacking is ImmutableList<JsonAny> items)
+            {
+                return new JsonArrayEnumerator(items);
+            }
+
+            if (this.jsonElementBacking.ValueKind == JsonValueKind.Array)
+            {
+                return new JsonArrayEnumerator(this.jsonElementBacking);
+            }
+
+            return default;
         }
 
         /// <inheritdoc/>
         public bool TryGetProperty(string name, out JsonAny value)
         {
-            return this.AsObject.TryGetProperty(name, out value);
+            if (this.objectBacking is ImmutableDictionary<string, JsonAny> properties)
+            {
+                return properties.TryGetValue(name, out value);
+            }
+
+            if (this.jsonElementBacking.ValueKind == JsonValueKind.Object)
+            {
+                if (this.jsonElementBacking.TryGetProperty(name, out JsonElement jsonElement))
+                {
+                    value = new JsonAny(jsonElement);
+                    return true;
+                }
+            }
+
+            value = default;
+            return false;
         }
 
         /// <inheritdoc/>
         public bool TryGetProperty(ReadOnlySpan<char> name, out JsonAny value)
         {
-            return this.AsObject.TryGetProperty(name, out value);
+            if (this.objectBacking is ImmutableDictionary<string, JsonAny> properties)
+            {
+                return properties.TryGetValue(name.ToString(), out value);
+            }
+
+            if (this.jsonElementBacking.ValueKind == JsonValueKind.Object)
+            {
+                if (this.jsonElementBacking.TryGetProperty(name, out JsonElement jsonElement))
+                {
+                    value = new JsonAny(jsonElement);
+                    return true;
+                }
+            }
+
+            value = default;
+            return false;
         }
 
         /// <inheritdoc/>
         public bool TryGetProperty(ReadOnlySpan<byte> utf8name, out JsonAny value)
         {
-            return this.AsObject.TryGetProperty(utf8name, out value);
+            if (this.objectBacking is ImmutableDictionary<string, JsonAny> properties)
+            {
+                return properties.TryGetValue(Encoding.UTF8.GetString(utf8name), out value);
```

Comment:
Yes; this was originally prototyped using Memory<char> until I twigged that it didn't support a HashCode based on the contents of the array!

The optimisation would be to build a hashable string-like type just for the keys to that dictionary.

Because you cannot (yet?) get `JsonElement` property names as UTF8 strings in any way (except by serialization of the whole property) you still have to do a conversion from UTF8. However, if this was converted into some UTF8-string-like-thing we could avoid allocating that string.